proposals/tf-operator-design-v1alpha2.md

```diff
@@ -0,0 +1,435 @@
+    + We use a postfix `uid` to make each object name unique.
+    + Then set these objects' `OwnerReferences` to this TFJob object.
+- Listen on pods/services via `podInformer` and `serviceInformer`.
+    + On pod created/updated/deleted, get TFJob object by parsing `OwnerReferences`, set the `TFJob.Status` as defined above according to the whole TF cluster state.
+    + Update the `TFJob.Status.Condition` if needed.
+- Terminate/Delete the TFJob object if every pod is completed (or leave pod phase as `Succeeded`).
+    + This maybe be lead to logs and model checkpoint files unreachable.
+
+### Reconciler
+
+More than that, we should provide a `Reconciler` mechanism to reconcile observed and desired states and repair discrepancies as a double check for `Event-Driven` mechanism.
+
+Here is `configuration` of tf-operator:
+
+```go
+// TFControllerConfiguration contains configuration of tf-operator.
+// DefaultTimerConfig is the suggested tf-operator configuration for production.
+type TFControllerConfiguration struct {
+    // ReconcilerSyncLoopPeriod is the amount of time the reconciler sync states loop
+    // wait between two reconciler sync.
+    // It is set to 15 sec by default.
+    // TODO(cph): maybe we can let it grows by multiple in the future
+    // and up to 5 minutes to reduce idle loop.
+    // e.g. 15s, 30s, 60s, 120s...
+    ReconcilerSyncLoopPeriod metav1.Duration
+}
+
+// DefaultTFControllerConfiguration is the suggested tf-operator configuration for production.
+var DefaultTFControllerConfiguration TFControllerConfiguration = TFControllerConfiguration{
+    ReconcilerSyncLoopPeriod: 15 * time.Second,
+}
+```
+
+Reconciler use a `ReconcilerSyncLoopPeriod` to determine whether we should call this reconciler or ignore it. We should leave a record `LastReconcileTime` in TFJob object of course:
+
+```go
+    // Represents last time when the TFJob was reconciled. It is not guaranteed to
+    // be set in happens-before order across separate operations.
+    // It is represented in RFC3339 form and is in UTC.
+    LastReconcileTime *metav1.Time `json:"lastReconcileTime,omitempty"`
+```
+
+As `tfJobImformer` provides a forcing resync mechanism by calling `UpdateFunc` which defined in `ResourceEventHandlerFuncs` periodically. We can call the reconciler in this function:
+
+- UpdateFunc return a TFJob object periodically.
+- Check `LastReconcileTime` to determine whether we should trigger a reconciler call.
+- `tf-operator` will list all pods/services which related to this TFJob.
+  + Compare the current state to the spec of this TFJob.
+  + Try to recovery the failed pod/service to make the training healthy.
+      + Error handing is described below.
+- Update the status of this TFJob.
+- TODO: we should call this reconciler with an exponential back-off delay (15s, 30s, 60s …) capped at 5 minutes.
+
+### Error Handling
+
+To make the system robust, the tf-operator should be able to locally and automatically recover from errors. 
```

Comment:
What about retryable vs. permanent errors? The current implementation uses exit codes to distinguish between retryable and permanent errors? Should we continue to use that?

Reply:
Hi, I think this is open to discuss. IMHO, separate retryable and permanent errors maybe not easy, so let user to decide this maybe better.

Reply:
So I agree user should decide whether an error is retryable or not, but that needs to be communicated to the TFJob operator so that it can determine what to do.

For example, here are two different cases where TF could be terminated

1. The VM running the pod is terminated or becomes unhealthy 
1. TF exits with an exception because user tries to read a file that doesn't exist.

My claim is that in the first case TF operator should restart TF since its a retryable error but in second case the error is most likely permanent (file is unlikely to exist if we keep retrying) so the job should fail.

Exit codes provide a mechanism for distinguishing different types of errors. User has some control over the exit code. For example they can wrap their code in a try/catch block and turn different exceptions into exit codes corresponding to retryable or permanent errors as needed.

This isn't great but its fairly straightforward.

Reply:
Hi, 

- In case 1, AFAIK this case will create a node `NotReady` condition, kubernetes will re-schedule all pods which running on this node to another node after a while. It is handled automatically by kubernetes system, not controlled by any user defined config (e.g. RestartPolicy or exit code).

- In case 2, this is a error in container, so let use handle this error maybe better, by setting RestartPolicy.

Reply:
In case 1, I don't think K8s will reschedule the pods. Pods aren't durable
https://kubernetes.io/docs/concepts/workloads/pods/pod/#durability-of-pods-or-lack-thereof
I think its the controller that creates a new pod but in our case we are creating/managing the pods ourselves.

In case 2. I don't think we can do this with the built in RestartPolicy because the only options are Never or Always and I don't think that's flexible enough.